enhancements/cloud-integration/credential-rotation.md

```diff
@@ -0,0 +1,183 @@
+---
+title: cloud-credential-rotation
+authors:
+  - "@dgoodwin"
+reviewers:
+  - "@joelddiaz"
+  - "@abhinavdahiya"
+  - "@jsafrane"
+  - "@dmage"
+  - "@enxebre"
+  - "@MaysaMacedo"
+approvers:
+  - "@sdodson"
+  - "@derekwaynecarr"
+creation-date: 2020-10-06
+last-updated: 2020-10-08
+status: provisional
+---
+
+# Cloud Credential Rotation
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [X] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Implement automatic credentials rotation in the cloud-credential-operator.
+Allow users to configure a global expiry duration for their cloud credentials config,
+after which the CCO will rotate new credentials in, and cleanup the old after a
+suitable period of time.
+
+## Motivation
+
+Customers often have security policies that prohibit the use of long lived
+cloud credentials. This enhancement provides a solution to a frequently
+requested feature, will improve OpenShift security overall, and help meet the
+requirements of more customers and thus generate more sales.
+
+### Goals
+
+  1. Support a global expiry time for CredentialsRequest credentials, defined in the CloudCredential config CR.
+  1. Rotate expired credentials when expiry time has been reached.
+  1. Preserve previous credentials for a period of time to ensure operators remain fully functional.
+  1. All operators using CredentialsRequests able to successfully handle a Secret modification with new credentials without going degraded.
+
+### Non-Goals
+
+  1. Mint mode only, not applicable to Passthrough or Manual mode. (although the operator support for credentials changing will be universally useful in both of these modes for customers manually rotating their own credentials)
+  1. AWS STS Token support. This is separate from this effort and not applicable to all mint capable cloud providers.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I would like to be able to define a credentials
```

Comment:
This is different from our other rotation apis, like the kube-apiserver-operator which
- picks a rotation interval
- doesn't allow user to configure this, only using an unsupported mechanism for we


Why would we allow users to configure this in the first place? Since OpenShift goal is to be always secure (we want to be production ready from get go) why would we openly allow a less secure configuration?

I think it's best for us to move everyone to always rotating temporary creds, where we cannot rotate permanent creds pick a good rotation period .. if the user doesnt like our impl let them take over completely. At least that's what I think is best ..

Reply:
I think I could get behind making rotation the default. Would you do the same for an upgraded cluster or only for new installs?

I'm also open to not making the interval configurable, and just picking a default internally hard coded. 

Will check with our PM to see if she has any objections but I think this seems reasonable.

Reply:
A little worried the rotation interval might need to be configurable, it's a *very* small thing to push someone to manual mode over if they had more restrictive lifetime requirements than we hard coded.